mirai-core/src/commonMain/kotlin/network/net/impl/netty/NetyNetworkHandlerFactory.kt

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2019-2021 Mamoe Technologies and contributors.
+ *
+ *  此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
+ *  Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
+ *
+ *  https://github.com/mamoe/mirai/blob/master/LICENSE
+ */
+
+package net.mamoe.mirai.internal.network.net.impl.netty
+
+import io.netty.channel.ChannelHandlerContext
+import io.netty.channel.ChannelInboundHandlerAdapter
+import kotlinx.coroutines.CompletableDeferred
+import net.mamoe.mirai.internal.network.net.NetworkHandlerContext
+import net.mamoe.mirai.internal.network.net.NetworkHandlerFactory
+import java.net.SocketAddress
+
+internal object NettyNetworkHandlerFactory : NetworkHandlerFactory<NettyNetworkHandler> {
+    override fun create(context: NetworkHandlerContext, address: SocketAddress): NettyNetworkHandler {
+        return NettyNetworkHandler(context, address)
+    }
+
+}
+
+internal class RetrieveContextInboundHandler : ChannelInboundHandlerAdapter() {
+    val instance: CompletableDeferred<ChannelHandlerContext> = CompletableDeferred()
+
+    override fun channelActive(ctx: ChannelHandlerContext) {
+        if (!instance.isCompleted) instance.complete(ctx)
+    }
+}
+
+internal object PacketDecodeInboundHandler : ChannelInboundHandlerAdapter() {
```

Comment:
`object` 在 Netty 中会有 `Sharable` 的问题
```java
public abstract class ChannelHandlerAdapter implements ChannelHandler {

    // Not using volatile because it's used only for a sanity check.
    boolean added;

    /**
     * Throws {@link IllegalStateException} if {@link ChannelHandlerAdapter#isSharable()} returns {@code true}
     */
    protected void ensureNotSharable() {
        if (isSharable()) {
            throw new IllegalStateException("ChannelHandler " + getClass().getName() + " is not allowed to be shared");
        }
    }

    /**
     * Return {@code true} if the implementation is {@link Sharable} and so can be added
     * to different {@link ChannelPipeline}s.
     */
    public boolean isSharable() {
        /**
         * Cache the result of {@link Sharable} annotation detection to workaround a condition. We use a
         * {@link ThreadLocal} and {@link WeakHashMap} to eliminate the volatile write/reads. Using different
         * {@link WeakHashMap} instances per {@link Thread} is good enough for us and the number of
         * {@link Thread}s are quite limited anyway.
         *
         * See <a href="https://github.com/netty/netty/issues/2289">#2289</a>.
         */
        Class<?> clazz = getClass();
        Map<Class<?>, Boolean> cache = InternalThreadLocalMap.get().handlerSharableCache();
        Boolean sharable = cache.get(clazz);
        if (sharable == null) {
            sharable = clazz.isAnnotationPresent(Sharable.class);
            cache.put(clazz, sharable);
        }
        return sharable;
    }
```